### PR TITLE
speeding up sha3

### PR DESCRIPTION
**Pull request by karthikbhargavan**
## Proposed changes

Step 1: avoiding a state copy in chi. 
Step 2: removing calls to salloc1 and replacing them with create.
Step 3: marking local constant arrays as "private"

Of these, we should revisit Step 2 when we have a systematic approach towards zeroing out stack-allocated values.
In the current code, almost all the state is in registers so the value of memzero is unclear.

The result of these changes is a 200% perf improvement when using clang on Linux and Mac OS. 
With GCC, there is a more modest performance improvement of 25-30%.
We will try to reconcile GCC and clang performance over time.

## Types of changes

What types of changes does your code introduce to HACL*?
_Put an `x` in the boxes that apply_

- [ ] Proof maintenance (non-breaking change which fixes a proof regression)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New algorithm or feature (non-breaking change which adds functionality)
- [X] Improved performance (fix or feature that would improve performance of some algorithm on some platform)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [X] I have read the [CODE_OF_CONDUCT.md](https://github.com/hacl-star/hacl-star/blob/main/CODE_OF_CONDUCT.md) doc
- [X] I have read and agree to submit my changes under the [LICENSE](https://github.com/hacl-star/hacl-star/blob/main/LICENSE)
- [ ] I have added necessary documentation (if appropriate)
- [ ] I have edited [CHANGES.md](https://github.com/hacl-star/hacl-star/blob/main/CHANGES.md) (if appropriate)

## Further comments

This a series of optimizations that close the performance gap between our SHA3 implementation and other C implementations. There remains a gap between this code and implementations that use AVX2 SIMD instruction.

### CODE REVIEW

**Review by polubelova — APPROVED** [review]
looks great, thanks!

File: dist/gcc-compatible/Hacl_SHA3.h
Comment (msprotz) [inline dist/gcc-compatible/Hacl_SHA3.h:38]: that file is not bundled behind an interface (see my comment in the other PR as to why), so you probably would benefit from adding `private` qualifiers in state_permute, loadState and storeState

File: dist/gcc-compatible/internal/Hacl_SHA3.h
Comment (msprotz) [inline dist/gcc-compatible/internal/Hacl_SHA3.h:1]: you might want to wait for my other PR to land first, as it will make this file disappear, I think

File: specs/lemmas/Spec.SHA3.Equivalence.fst
Comment (msprotz) [reply specs/lemmas/Spec.SHA3.Equivalence.fst:1]: maybe, there's also a precedent for putting a lot of spec equivalence lemmas in here

**Review by msprotz — CHANGES_REQUESTED** [review]
Looks good minus my comment about marking a couple more things private. Let's wait for #780 to land first, as I think it'll avoid creating a new file

File: specs/lemmas/Spec.SHA3.Equivalence.fst
Comment (karthikbhargavan) [reply specs/lemmas/Spec.SHA3.Equivalence.fst:1]: I don't mind either way. I could turn Spec.SHA3.Equivalence into code/sha3/Hacl.Spec.Sha3 or I could leave it here.

File: dist/gcc-compatible/Hacl_SHA3.h
Comment (karthikbhargavan) [reply dist/gcc-compatible/Hacl_SHA3.h:38]: I think permute is called from the streaming APIs and it complains when I turn it private. I guess it would be better to make these functions private and instead expose "absorb_inner" which functions as "update-1-block" for SHA3?

File: dist/gcc-compatible/internal/Hacl_SHA3.h
Comment (karthikbhargavan) [reply dist/gcc-compatible/internal/Hacl_SHA3.h:1]: Oh year, I was wondering why this file was being generated. Happy to have it go away.

File: dist/gcc-compatible/Hacl_SHA3.h
Comment (msprotz) [reply dist/gcc-compatible/Hacl_SHA3.h:38]: curious why streaming calls these functions... wondering if something is inlined too aggressively

Comment (msprotz) [reply dist/gcc-compatible/Hacl_SHA3.h:38]: yeah so the reason is because streaming calls absorb_inner to do the loop over a multiple of a number of blocks, and that one is inline_for_extraction noextract, which means the implementation of SHA3 is replicated in the streaming version and hence generates cross-translation unit calls to permute

my suggested fix is to remove inline_for_extraction noextract from `absorb_inner`, and wait for my other PR to land so that you won't have a performance hit (because as streaming_sha3 and sha3 are merged together into a single file, `absorb_inner` will be marked as `static` and hopefully will happily optimized just as well)

alternatively, you could do `let absorb_inner_non_inline = absorb_inner`, then call `absorb_inner_non_inline` from `Hacl.Hash.SHA3`, which will remove the copy of `absorb_inner` inlined everything in Hacl.Streaming.SHA3, and will allow you to mark permute as private

**Comment by karthikbhargavan** [comment]
@msprotz is there anything other than waiting for the other PR to land?

**Review by msprotz — APPROVED** [review]
No that's good for me. Just waiting on Franziskus to approve the other PR. It might impact hacl-packages so I want his green light.

File: dist/gcc-compatible/internal/Hacl_SHA3.h
Comment (msprotz) [reply dist/gcc-compatible/internal/Hacl_SHA3.h:1]: yeah the file is still there because EverCrypt.Hash repeatedly calls `absorb_inner` in the update_multi loop, and this is a call across translation units, so it generates an internal header

(I forgot about that caller)

Comment (msprotz) [reply dist/gcc-compatible/internal/Hacl_SHA3.h:1]: I'll merge this PR, but if the internal header is bad, and you need loadState to be marked as `static` for perf reasons, open up an issue and I'll fix it

**Pull request merged**
